File: code/python/1-100/11_maxArea.py
# author: Xiname
# -*- coding: utf-8 -*-
# @Time : 2023/12/21 11:40
# @File ：11_maxArea.py
# @IDE ：PyCharm
from typing import List


class Solution:
    def maxArea(self, height: List[int]) -> int:
        # 初始化两个指针 i=0, j=len(height)-1
        i, j = 0, len(height) - 1
        maxArea = 0
        while i != j:

            if height[i] < height[j]:
                min_height = height[i]
                # print('i = {}, j = {}'.format(i, j))
                i += 1
            else:
                min_height = height[j]
                # print('i = {}, j = {}'.format(i, j))
                j -= 1
            # 这里长度变成 j - i + 1 是因为前面会多减一个1，造成结果有问题！！！
            maxArea = max(maxArea, (j - i + 1) * min_height)
            # print('maxArea = {},min_height = {}'.format(maxArea, min_height), )
        return maxArea


if __name__ == '__main__':
    s = Solution()
    # print(s.maxArea([1,1]))
    print(s.maxArea([1, 8, 6, 2, 5, 4, 8, 3, 7]))

File: code/cpp/1-100/11_maxArea.cpp
#include <bits/stdc++.h>
using namespace std;
//题目
//给定一个长度为 n 的整数数组 height 。有 n 条垂线，第 i 条线的两个端点是(i, 0) 和(i, height[i]) 。
//找出其中的两条线，使得它们与 x 轴共同构成的容器可以容纳最多的水。
//返回容器可以储存的最大水量。

class Solution {
public:
    int maxArea(vector<int>& height) {
        int nLeft = 0, nRight = height.size() - 1;
        int nMaxArea = 0;
        while (nLeft != nRight)
        {
            nMaxArea = max(nMaxArea, min(height[nLeft], height[nRight]) * (nRight - nLeft));
            if (height[nLeft] < height[nRight])
            {
                ++nLeft;
            }
            else
            {
                --nRight;
            }
        }
        nMaxArea = max(nMaxArea, min(height[nLeft], height[nRight]) * (nRight - nLeft));
        return nMaxArea;
    }
};

//int main()
//{
//    Solution S;
//    //vector<int> height = { 1, 8, 6, 2, 5, 4, 8, 3, 7 };
//    vector<int> height = { 1,1 ,1};
//    cout << S.maxArea(height);
//	return 0;
//}